cuda/solver/common_trs_kernels.cuh

```diff
@@ -474,9 +502,15 @@ __global__ void sptrsv_naive_caching_kernel(
                 val = load_relaxed_shared(x_s + dependency_shid))) {
             }
         } else {
+            // We must check the original type because is_nan_exact one kind of
+            // quiet_NaN from gko::nan. Different precision may use the
+            // different kind of quiet_NaN and the casting result is not clear
+            // because the quiet_NaN is not specified by IEEE.
+            auto output_val = zero<ValueType>();
             while (is_nan_exact(
-                val = load_relaxed(x + dependency * x_stride + rhs))) {
+                output_val = load_relaxed(x + dependency * x_stride + rhs))) {
```

Comment:
Since the conversions are the issue here, I would propose a slightly different approach:
1. Make sure that `x` stores the bit-wise exact nan value of `SharedValueType`, via `memcpy`.
2. Use a function `bitwise_equal(a, b)`, where `a` is the loaded value and `b` the expected nan value.

I think this could make it a bit clearer what is happening, and not use the `gko<half|bfloat>` types internally.

Reply:
Maybe I get the first point wrong. x is stored exact nan value because we fill it by `gko::nan<ValueType>` and x_shared is stored by `gko::nan<SharedValueType>`. We do not have the casting during assignment.

I think `bitwise_equal(a, b)` indeed makes the comparison value more clear. (or we can make it more clear/consistent what the exact nan among different backend/compiler)
Unfortunately, we do not have the `host_type` map currently to do `bitwise_eqaul(a, nan<host_type<T>>)` and I would not like to introduce another type mapping here.

Reply:
You're right, I guess it was getting a bit late for me. Could an alternative be to fill the `x` with `gko::nan<device_type<ValueType>>` already? I will still approve this, just curious.